src/status_im/contexts/keycard/unblock/events.cljs

```diff
@@ -2,86 +2,89 @@
+                     (rf/dispatch [:keycard/unblock.generate-and-load-key])))}]))
 
 ;; second stage, init card with pin
+;; there is no way to compare instance-uid for reseted card
 (defn init-with-pin
-  []
+  [masked-pin]
   (rf/dispatch
-   [:keycard/connect.next-stage
-    {:on-error (fn [error]
-                 (if (= error :keycard/error.keycard-empty)
-                   (rf/dispatch [:keycard/unblock.init-card-or-import-keys])
-                   (do
-                     (rf/dispatch [:navigate-back])
-                     (rf/dispatch [:keycard/on-application-info-error error]))))}]))
+   [:keycard/connect
+    {:next-stage? true
+     :on-success
+     (fn [{:keys [initialized? has-master-key?]}]
+       (if has-master-key?
+         (show-different-card)
+         (if initialized?
+           (import-keys)
+           (rf/dispatch [:keycard/init-card
+                         {:pin        (security/safe-unmask-data masked-pin)
+                          :on-success import-keys}]))))}]))
 
 ;; first stage, reset card
 (rf/reg-event-fx :keycard/unblock
  (fn [{:keys [db]}]
-   {:fx [[:dispatch
-          [:keycard/connect
-           {:key-uid  (get-in db [:keycard :application-info :key-uid])
-            :on-error (fn [error]
-                        (cond
-                          (or (= error :keycard/error.keycard-frozen)
-                              (= error :keycard/error.keycard-locked))
-                          (rf/dispatch [:keycard/factory-reset {:on-success init-with-pin}])
-
-                          (= error :keycard/error.keycard-empty)
-                          (init-with-pin)
-
-                          :else
-                          (do
-                            (rf/dispatch [:navigate-back])
-                            (if (= error :keycard/error.keycard-wrong-profile)
-                              (do
-                                (rf/dispatch [:keycard/disconnect])
-                                (rf/dispatch [:open-modal :screen/keycard.different-card]))
-                              (rf/dispatch [:keycard/on-application-info-error error])))))}]]]}))
+   (let [masked-pin (get-in db [:keycard :unblock :masked-pin])]
+     {:fx [[:dispatch
+            [:keycard/connect
+             {:on-success
+              (fn [{:keys [initialized?]}]
+                (if initialized?
+                  (show-different-card)
+                  (init-with-pin masked-pin)))
+              :on-error
+              (fn [error {:keys [instance-uid]}]
+                (if (not= instance-uid (get-in db [:keycard :unblock :instance-uid]))
+                  (show-different-card)
+                  (if (or (= error :keycard/error.keycard-frozen)
+                          (= error :keycard/error.keycard-locked))
+                    (rf/dispatch [:keycard/factory-reset {:on-success #(init-with-pin masked-pin)}])
+                    (rf/dispatch [:keycard/on-application-info-error error]))))}]]]})))
```

Comment:
hi @flexsurfer, Thank you very much for you great work. I Will review PR tomorrow.

> Implement on-complete event instead of on-error and on-succe

It seems we still have `on-success` and `on-error`. Is the PR still a work in progress, or will this be handled separately? If will be handled separately, should the PR title be updated?

Reply:
hey @Parveshdhull thank you, i'll notify you when it's ready for review, still WIP